code/src/nuvla/job/actions/coe_provision.py

```diff
@@ -0,0 +1,112 @@
+
+        coe_type = infra_service_coe['subtype']
+        if coe_type not in [COE_TYPE_SWARM, COE_TYPE_K8S]:
+            raise Exception(f'Unknown COE type: {coe_type}')
+
+        cloud_creds_id = infra_service_coe.get('management-credential')
+        coe_custer_params = copy.deepcopy(infra_service_coe.get('cluster-params', {}))
+        if 'ssh-keys' in coe_custer_params:
+            # Update ssh keys as IDs by their public key values.
+            ssh_pub_keys = []
+            for ssh_key_id in coe_custer_params['ssh-keys']:
+                # TODO: use search() to get them all at once.
+                ssh_key = self.api.get(ssh_key_id).data['public-key']
+                ssh_pub_keys.append(ssh_key)
+            coe_custer_params['ssh-keys'] = ssh_pub_keys
+
+        coe = connector_factory(docker_machine_connector, self.api,
+                                cloud_creds_id, infra_service_coe)
+
+        result = coe.start(coe_custer_params)
+        try:
+            self.job.set_progress(50)
+            self._register_coe(infra_service_coe, result)
+            self.job.set_progress(99)
+        except Exception as ex:
+            msg = f'Terminating deployed cluster. Failure in registration: {ex}'
+            logging.warning(msg)
+            stopped = coe.stop(nodes=result['nodes'])
+            raise ex
+
+        return 0
+
+    def _register_coe(self, infra_service_coe: dict, result: dict):
+
+        # Minimal ACL for COE credentials. Only owners can see it.
+        acl = {"acl": {"owners": infra_service_coe['acl']['owners']}}
+
+        coe_cred_data = {**result['creds'], **acl}
+        infra_service_coe_id = infra_service_coe['id']
+        self._set_coe_creds(coe_cred_data, infra_service_coe['subtype'],
+                            infra_service_coe.get('name'),
+                            infra_service_coe_id)
+
+        cluster_params = infra_service_coe.get('cluster-params', {})
+        if 'coe-manager-endpoint' in result:
+            cluster_params['coe-manager-endpoint'] = result['coe-manager-endpoint']
+        if 'join-tokens' in result:
+            cluster_params['join-tokens'] = result['join-tokens']
+        self.api.edit(infra_service_coe_id,
+                      {'endpoint': result['endpoint'],
+                       'state': 'STARTED',
+                       'nodes': result['nodes'],
+                       'cluster-params': cluster_params})
+
+    def _set_coe_creds(self, coe_cred_data, coe_type, cred_name,
+                       infra_service_coe_id):
+        if coe_type == COE_TYPE_SWARM:
+            coe_creds = CredentialDockerSwarm \
+                .build_template(coe_cred_data, infra_service_coe_id,
+                                cred_name,
+                                f'Credential for infrastructure service {cred_name}')
+        elif coe_type == COE_TYPE_K8S:
+            coe_creds = CredentialK8s \
+                .build_template(coe_cred_data, infra_service_coe_id,
```

Comment:
build_template is a newly defined method in credential helper of python-library. 
I think we need to release python-library 3.0.2 and update the requirement dependency in job-engine to avoid feature break at deployment

Reply:
Yes, this is the case.